GH-98388: Add tests for happy eyeballs and its internal workings

twisteroidambassador opened the pull request:
GH-98388

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

twisteroidambassador committed:
Add tests for happy eyeballs and its internal workings.

One of the new tests may fail intermittently due to #86296.

gvanrossum commented on Lib/test/test_asyncio/test_base_events.py:
```diff
@@ -145,6 +145,58 @@ def test_ipaddr_info_no_inet_pton(self, m_socket):
                                                    socket.IPPROTO_TCP))
 
+    def test_interleave_ipaddrs(self):
+        addrinfos = [
```
These tests would be a lot more readable if you gave each line a short name, so you could do e.g.
```py
A = ...
B = ...
...
addrinfos = [A, B, C, D, E, F, G, H]
expected = [A, E, B, F, C, G, D, H]
```
The definitions of A-H can be shared between the two tests.

twisteroidambassador replied:
Done. I gave them names SIX_A, FOUR_A, etc to distinguish between IPv6 and IPv4 addresses.

gvanrossum reviewed (changes requested):
I've added the "skip news" label because this is a test-only PR.

twisteroidambassador committed:
Correct typo in filename.

twisteroidambassador committed:
Improve readability for _interleave_addrinfos tests.

twisteroidambassador committed:
Rename _interleave_addrinfos tests to match the method being tested.

twisteroidambassador committed:
Made test_simultaneous_success_fail fail more consistently (I hope).

twisteroidambassador committed:
Replace `wait_for` with `wait`.

This solution is compatible with all Python versions, and should pass all tests.

kumaraditya303 reviewed (changes requested):
Once https://github.com/python/cpython/pull/98518 lands, please remove the implementation changes and only add tests.

arhadthedev committed:
Delete staggered.py

arhadthedev committed:
Merge branch 'main' into happy_eyeballs_checkup

arhadthedev committed:
Restore staggered.py from `main` branch

arhadthedev committed:
Address patchcheck report

gvanrossum committed:
Merge branch 'main' into happy_eyeballs_checkup

gvanrossum commented on Lib/test/test_asyncio/test_base_events.py:
```diff
@@ -145,6 +145,27 @@ def test_ipaddr_info_no_inet_pton(self, m_socket):
+        FOUR_C = (socket.AF_INET, 0, 0, '', ('192.0.2.3', 7))
+        FOUR_D = (socket.AF_INET, 0, 0, '', ('192.0.2.4', 8))
+
+        addrinfos = [SIX_A, SIX_B, SIX_C, SIX_D, FOUR_A, FOUR_B, FOUR_C, FOUR_D]
```
Would it be a more thorough test if we mixed up the order a bit, e.g. like this?
```suggestion
        addrinfos = [SIX_A, SIX_B, SIX_C, FOUR_A, FOUR_B, FOUR_C, FOUR_D, SIX_D]
```

willingc replied:
LGTM @gvanrossum.

gvanrossum commented on Lib/test/test_asyncio/test_staggered.py:
```diff
@@ -0,0 +1,115 @@
+        self.assertIsInstance(exceptions[1], RuntimeError)
+
+    async def test_simultaneous_success_fail(self):
+        # There's a potential race condition here:
+        # https://github.com/python/cpython/issues/86296
+        # As with any race condition, it can be difficult to reproduce.
+        # This test may not fail every time.
```
Please add a note to this comment that the race condition has been fixed in 3.12, e.g.
```suggestion
        # There's a potential race condition here (fixed in Python 3.12):
        # https://github.com/python/cpython/issues/86296
        # As with any race condition, it can be difficult to reproduce.
        # This test may not fail every time.
```

gvanrossum reviewed (commented):
This basically LGTM, I only have very minor nits.

@arhadthedev, would you take over this PR, since @twisteroidambassador seems to have disappeared? I think it would be good to land this and backport it to 3.12. At the very least you could review my suggestions, if you agree with everything I can merge.

@willingc Did you want to add anything?
